### DISCUSSION TITLE
Converting an existing product

### DISCUSSION THREAD

**Discussion by hammyau** [eclipse-tycho/tycho#418] [Q&A] [answered]
Hi All,
I have an existing RCP project to maintain. Lucky me :-) I am looking to switch it to using maven pomless tycho (2.5) from its existing (ant script somewhere) build mechanism.

I have most of the pieces building but not the main product. It can't find the source.

Is there a good and straight forward explanation of how to set up an RCP product? I think it is a naming thing.
I have read many pages of unclear documentation.
I have Vogella's examples building and running. But do not understand enough how these hang together and its structure is different to mine. It has Bundles, Features directories etc mine is just projects under a root level.

I have created a releng directory to try to define the target. But that is all a bit of a mystery to me.

I have even resorted to looking through the plugin source to see if I can figure it out.
Incidentally how do you debug a maven plugin?

Should say this project dates back to the Luna days.
It all builds and runs through the 2021-09 IDE execution. But I want a command line build build.

Thanks in advance.

Ian C

**Comment by mickaelistria** [comment]
What specific error message do you get?

**Reply by hammyau** [reply]
Thanks for the reply Mickael. I have had so many different passes through this thing it is hard to specify a given error message. My latest actually got through install with everything green and produced and executable jar. Which came as something of a surprise. I am trying to make an RCP product.

The lack of documentation on this thing really makes it difficult.

**Comment by laeubi** [comment]
Have you used [ant4eclipse](http://www.ant4eclipse.org/) before? Or custom ant/pde-build like?

Anyways it is rather straight forward to setup a pomless-build if you follwo the standard structure:

1. have a root folder where you "configuration pom" is located
2. a folder `.mvn` where you configure the pomless extension in the `extensions.xml`
3. a folder `plugins` (or `bundles`) where you place your projects
4. a folder `features` where you place your features
5. a folder `products` where you place your products

A releng folder is no longer necessary as you can configure the target as a file-reference, for self-contained builds this is the recommended approach. If anything is not working, you should share the error message and at best a public example.

Beside that if you require commercial support, my company offers [open source software support](https://xn--lubisoft-0za.gmbh/en/#opensource) (including Tycho), feel free to contact me.

**Reply by hammyau** [reply]
Thanks Christoph. Where is that "standard structure" documented?
I am learning quite a bit about this stuff. But in so many little fragments.

I may try an iteration moving the parts around into the folders you suggest. Is there any documentation on the target definition?

**Reply by hammyau** [reply]
Ah I see something on your site. I'll see what I can make of that.

**Comment by laeubi** [comment]
If your coming from ant maven might feels a bit alien... In general the documentation is part of the different mojos for the target platform you can find it here: https://www.eclipse.org/tycho/sitedocs/target-platform-configuration/target-platform-configuration-mojo.html

The best you can do is create a little github project and try to setup a simple project setup (e.g. one bundle + one target file + one feature + one product) so you can share it for any problem you come across.

Later on its just a matter of simply adding more content to this. There are some (a bit outdated) outlines here: https://wiki.eclipse.org/Tycho/pomless most of this is still valid for current tycho-pomless build.

Sadly I haven't had yet the opportunity to migrate (and upgrade) it to github, so if you like to help out a first step might be to copy the current content to https://github.com/eclipse/tycho/wiki/Tycho-Pomless adjust the formating and then ask specific questions what is unclear or needs more infor so we can improve the documentation together. Its always good to have an outside view to find missing pieces.

**Reply by hammyau** [reply]
Thanks I'll take a look and update the wiki once I understand more.

**Reply by hammyau** [reply]
Hi, I have been bashing away at trying to understand what is going on here. I have even built the tycho repo in the hope that I can trace through what is happening.  That I have yet to try.... wish me luck. Will mvnDebug work.... we shall see.

I still don't understand some of the fundamental details here.
Like the recommended layout/structure and how it relates to the content.
I found a couple of interesting articles from Dirk Fauth http://blog.vogella.com/2019/11/25/pom-less-tycho-enhanced re the structure. They were useful.
I also see in the code that the use of "bundles, products" etc is defined.

Does that mean they do not need to be specified in the parent pom as modules? It seems they do still need to be there.

I did as suggested and created a trivial RCP app as per the Eclipse tutorial and then tried to get it to build via maven.
Manually moving the code to a bundles (and a products in another iteration)  sub directory.
I have a parent pom and the .mvn extensions (tycho version 2.5.0)

No joy. mvn install can see the items and in fact generates a nice green SUCCESS. But nothing is built.

A fundamental part of the game seems to revolve around the target file. 
Can you explain where it fits in. Do I still need to define one to get maven to drag in the needed eclipse equinox jars. It seems so since I am getting 

`Missing requirement: rcpapp 0.0.0 requires 'org.eclipse.equinox.p2.iu; org.eclipse.equinox.executable.feature.group 0.0.0' but it could not be found`

A target file is generated as part of the pomless extensions.
(As an aside how, for debug purposes, can you get the generated pom and target to not be deleted at the end of the run?)

If I need one where do I put it? You mentioned no longer needing a releng. If so how do you manage the target?

I am trying to understand the data flow here and if I can I will document it in the wiki (or wherever is appropriate). With some pretty pictures... I think in pictures more that straight text.

Looking forward to clearing my head of this confusion, being able to get a reliable build, and get on with the code I am supposed to be writing :-)

I think I am close... but missing a couple of vital pieces of the puzzle.

**Reply by laeubi** [reply]
Its a bit hard to understand what you have already tried and what was causing problems, as you already have a gihub acount you might want to [create a new repository](https://github.com/new) in your account an upload the code that way its a bit easier to see whats might be wrong.

> A fundamental part of the game seems to revolve around the target file.

Actually tycho can work without any target *but* that obviously is very uncommon as one in most cases like to not build up from the ground but using some kind of basis (e.g. eclipse, equinox or alike), so the very first quest is how does it currently work in eclipse? Are you using a target file, are you using the eclipse platform as an (implicit) target, or do you manage your dependecies some other way ([ivy](https://ant.apache.org/ivy/) maybe?

> Does that mean they do not need to be specified in the parent pom as modules?

You specify them as modules but you don't need poms inside the bundles/features/products folders anymore.

> But nothing is built

If the build succeeds at least *something* is build but as mentioned above you better upload what you have tried here so we can take a look.

> it seems so since I am getting

This indicates that you build a product with native launcher items but the native features are missing make sure your target contains the
`org.eclipse.equinox.executable.feature.group`

> A target file is generated as part of the pomless extensions.

I'm not aware of such a feature, a target file is actually an input of the build not the result of a build.

> I think I am close... but missing a couple of vital pieces of the puzzle.

It might be better to open more specific Q&A discussions e.g.: "How do I configure a t target file in a pomless build" or something alike (again here a code sample describing the problem could enable better answers).

**Reply by hammyau** [reply]
Thanks. Will do. Something is just not getting through my thick skull with this stuff.

**Reply by hammyau** [reply]
>     A target file is generated as part of the pomless extensions.
> 
> I'm not aware of such a feature, a target file is actually an input of the build not the result of a build.

Sorry my bad description getting confused with the 
`Computing target platform for MavenProject`
message

**Reply by hammyau** [reply]
Hi Christoph - public repository that does not work. Is hammyau/mytycho#1

**Comment by laeubi** [comment, marked as answer]
I have left two comments as new issues:

- https://github.com/hammyau/mytycho/issues/2
- https://github.com/hammyau/mytycho/issues/3

HTH

**Reply by hammyau** [reply]
Excellent! Thank you Christoph... 
That almost gets me there.
Now I am getting some unresolved dependencies which I am in the process of finding

Some of which appear as warnings.
This seems to be more of an Error than a warning? Since it stops the generation.

**Reply by laeubi** [reply]
You can post the log output as an issue in your github project if you like then it is a bit easier to decide what they mean.

**Reply by hammyau** [reply]
Eureka - it builds and runs.

**Comment by hammyau** [comment]
I was getting messages of the form

[WARNING] Mirror tool: Problems resolving provisioning plan.: [Unable to satisfy dependency from rcpapp 1.0.0.202112140718 to org.eclipse.equinox.p2.iu; org.slf4j.api [1.7.30.v20200204-2150,1.7.30.v20200204-2150].]

Which resulted in failures later. 
I edited the target-platform to get some of the dependencies from the correct repository. And also removed a number from the product file. 

Eclipse seems to add a whole bunch of things that are not required. Like the m2e entries. 
I seem to remember reading somewhere that the development dependencies are dragged in?

**Comment by hammyau** [comment]
Now equipped with my new knowledge I will try to apply it to my actual application.

**Reply by laeubi** [reply]
Good luck :+1:

**Comment by hammyau** [comment]
Well. I have my project building and running.
It has been a struggle due to not really understanding what I am doing. Read hacked it into submission.
The result seems to work but I am sure I have done many things incorrectly.

I will try to document my experiences in the wiki. They may help others who have to do the same thing.
And those more knowledgeable can comment to clarify.
